File: tckit/utils/tc_file_parser.py
"""tc_file_parser — XML/CDATA utilities for .TcPOU, .TcGVL, and .TcDUT files.

Shared utility used by reader and doc-generator adapters. Lives under
``tckit/utils/`` so adapters can depend on it without violating the
adapter-isolation rule (no adapter-to-adapter imports).

Stdlib only: xml.etree.ElementTree, re, pathlib.
No third-party dependencies.

TwinCAT file structure:
  .TcPOU — XML root contains either:
              <POU>  — for FUNCTION_BLOCK, FUNCTION, PROGRAM
              <Itf>  — for INTERFACE
            Both have optional <Method>, <Action>, <Property> children.
            <Property> elements contain <Get> and/or <Set> sub-elements.
  .TcGVL  — XML root contains a <GVL> element with a <Declaration> CDATA.
  .TcDUT  — XML root contains a <DUT> element with a <Declaration> CDATA
            (STRUCT, ENUM, UNION, or TYPE alias definitions).
"""

import re
import xml.etree.ElementTree as ET
from pathlib import Path

# ---------------------------------------------------------------------------
# Low-level XML helpers
# ---------------------------------------------------------------------------


def parse_file(path: Path) -> ET.Element:
    """Parse an XML file and return the root element.

    Raises:
        ValueError: If the file cannot be parsed as XML.
    """
    try:
        tree = ET.parse(path)
    except ET.ParseError as exc:
        raise ValueError(f"XML parse error in {path}: {exc}") from exc
    return tree.getroot()


def get_cdata(element: ET.Element | None) -> str:
    """Return the text content of an element, stripped of leading/trailing whitespace.

    ElementTree doesn't distinguish CDATA from plain text — the content is
    already decoded. Returns empty string if element is None or has no text.
    """
    if element is None:
        return ""
    return (element.text or "").strip()


def get_declaration(element: ET.Element) -> str:
    """Extract Declaration CDATA from a <POU>, <Itf>, <Method>, <Action>, etc."""
    return get_cdata(element.find("Declaration"))


def get_st_body(element: ET.Element) -> str:
    """Extract ST implementation body from Implementation/ST CDATA."""
    impl = element.find("Implementation")
    if impl is None:
        return ""
    st = impl.find("ST")
    return get_cdata(st)


# ---------------------------------------------------------------------------
# POU / Interface type detection
# ---------------------------------------------------------------------------


def detect_pou_type(declaration: str, element_tag: str = "POU") -> str:
    """Detect POUType from the element tag and declaration text.

    <Itf> elements are always "interface".
    <POU> elements are detected from the first keyword in the declaration.
    Defaults to "function_block" if no keyword is found.
    """
    if element_tag.lower() == "itf":
        return "interface"
    text = declaration.upper()
    for keyword, pou_type in (
        ("FUNCTION_BLOCK", "function_block"),
        ("FUNCTION", "function"),
        ("PROGRAM", "program"),
        ("INTERFACE", "interface"),
    ):
        if keyword in text:
            return pou_type
    return "function_block"


# Access modifiers (TwinCAT IEC 61131-3 ext): may appear between the
# METHOD / PROPERTY keyword and the name. XAE writes them when the
# accessor is created via the automation interface (CreateChild vInfo
# carries the access string), so a declaration produced by add_property
# / add_method looks like ``PROPERTY PUBLIC Foo : LREAL`` rather than
# ``PROPERTY Foo : LREAL``. Both spellings must round-trip cleanly.
_ACCESS_MOD = r"(?:(?:PUBLIC|PRIVATE|PROTECTED|INTERNAL|FINAL|ABSTRACT)\s+)*"


def extract_method_return_type(declaration: str) -> str:
    """Extract return type from a METHOD declaration using regex.

    Matches:  METHOD [<access>...] MethodName : ReturnType
    Returns empty string if not found.
    """
    match = re.search(
        rf"METHOD\s+{_ACCESS_MOD}\w+\s*:\s*(\w+)",
        declaration,
        re.IGNORECASE,
    )
    return match.group(1) if match else ""


def extract_property_return_type(declaration: str) -> str:
    """Extract return type from a PROPERTY declaration using regex.

    Matches:  PROPERTY [<access>...] PropertyName : ReturnType
    Returns empty string if not found.
    """
    match = re.search(
        rf"PROPERTY\s+{_ACCESS_MOD}\w+\s*:\s*(\w+)",
        declaration,
        re.IGNORECASE,
    )
    return match.group(1) if match else ""


# Matches an implementation-only VAR block:
#   VAR  ... END_VAR        (method locals)
#   VAR_TEMP ... END_VAR    (per-call temporaries)
#   VAR CONSTANT ... END_VAR (internal constants)
# Each block opener must be the only token on its line so we don't accidentally
# match VAR_INPUT/VAR_OUTPUT/VAR_IN_OUT/VAR_INST, which are part of the API
# surface and must be preserved.
_LOCAL_VAR_BLOCK_RE = re.compile(
    r"^[ \t]*(?:VAR(?:[ \t]+CONSTANT)?|VAR_TEMP)[ \t]*\r?\n.*?^[ \t]*END_VAR[ \t]*\r?\n?",
    re.MULTILINE | re.DOTALL | re.IGNORECASE,
)


def strip_method_locals(declaration: str) -> str:
    """Strip implementation-only VAR blocks from a method declaration.

    Removes VAR (locals), VAR_TEMP, and VAR CONSTANT blocks while preserving
    VAR_INPUT, VAR_OUTPUT, VAR_IN_OUT, and VAR_INST (the API surface).

    Used when building MethodSignature payloads for ``get_pou_interface`` so
    the interface call doesn't carry implementation detail. ``get_pou_item``
    must NOT call this — callers asking for a method body need its locals.
    """
    return _LOCAL_VAR_BLOCK_RE.sub("", declaration).rstrip()


# ---------------------------------------------------------------------------
# File-level parsers
# ---------------------------------------------------------------------------


def parse_tcpou(path: Path) -> dict:
    """Parse a .TcPOU file and return a dict with POU/Interface structure.

    Handles both <POU> (function blocks, functions, programs) and
    <Itf> (interfaces) root elements.

    Returns:
        {
            "name": str,
            "pou_type": str,             # "function_block"|"function"|"program"|"interface"
            "declaration": str,
            "body": str,
            "methods": [{"name": str, "declaration": str, "body": str}],
            "actions":  [{"name": str, "declaration": str, "body": str}],
            "properties": [
                {
                    "name": str,
                    "declaration": str,  # PROPERTY header
                    "get": {"declaration": str, "body": str} | None,
                    "set": {"declaration": str, "body": str} | None,
                }
            ],
        }
    """
    root = parse_file(path)

    # TwinCAT stores POUs as <POU> and interfaces as <Itf>
    pou_el = root.find("POU")
    itf_el = root.find("Itf")
    container = pou_el if pou_el is not None else itf_el
    if container is None:
        raise ValueError(f"No <POU> or <Itf> element found in {path}")

    tag = container.tag  # "POU" or "Itf"
    name = container.get("Name", "")
    declaration = get_declaration(container)
    body = get_st_body(container)
    pou_type = detect_pou_type(declaration, tag)

    methods = []
    for method_el in container.findall("Method"):
        methods.append(
            {
                "name": method_el.get("Name", ""),
                "declaration": get_declaration(method_el),
                "body": get_st_body(method_el),
            }
        )

    actions = []
    for action_el in container.findall("Action"):
        actions.append(
            {
                "name": action_el.get("Name", ""),
                "declaration": get_declaration(action_el),
                "body": get_st_body(action_el),
            }
        )

    properties = []
    for prop_el in container.findall("Property"):
        prop_name = prop_el.get("Name", "")
        prop_decl = get_declaration(prop_el)

        get_el = prop_el.find("Get")
        set_el = prop_el.find("Set")

        prop_get = (
            {"declaration": get_declaration(get_el), "body": get_st_body(get_el)}
            if get_el is not None
            else None
        )
        prop_set = (
            {"declaration": get_declaration(set_el), "body": get_st_body(set_el)}
            if set_el is not None
            else None
        )

        properties.append(
            {
                "name": prop_name,
                "declaration": prop_decl,
                "get": prop_get,
                "set": prop_set,
            }
        )

    return {
        "name": name,
        "pou_type": pou_type,
        "declaration": declaration,
        "body": body,
        "methods": methods,
        "actions": actions,
        "properties": properties,
    }


def parse_tcgvl(path: Path) -> dict:
    """Parse a .TcGVL file and return a dict with GVL structure.

    Returns:
        {"name": str, "declaration": str}
    """
    root = parse_file(path)
    gvl_el = root.find("GVL")
    if gvl_el is None:
        raise ValueError(f"No <GVL> element found in {path}")

    return {"name": gvl_el.get("Name", ""), "declaration": get_declaration(gvl_el)}


def parse_tcdut(path: Path) -> dict:
    """Parse a .TcDUT file and return a dict with DUT structure.

    DUTs include STRUCT, ENUM, UNION, and TYPE alias definitions.

    Returns:
        {"name": str, "declaration": str}
    """
    root = parse_file(path)
    dut_el = root.find("DUT")
    if dut_el is None:
        raise ValueError(f"No <DUT> element found in {path}")

    return {"name": dut_el.get("Name", ""), "declaration": get_declaration(dut_el)}


# ---------------------------------------------------------------------------
# Project-file parsers (.plcproj / .tsproj / .TcTTO)
# ---------------------------------------------------------------------------


def _local(tag: str) -> str:
    """Strip XML namespace from a tag, e.g. '{ns}Project' -> 'Project'."""
    return tag.split("}", 1)[-1] if "}" in tag else tag


def _to_int(value: str | None) -> int | None:
    if value is None:
        return None
    try:
        return int(value.strip())
    except (ValueError, TypeError, AttributeError):
        return None


def _child_text(element: ET.Element, local_name: str) -> str:
    for child in element:
        if _local(child.tag) == local_name:
            return (child.text or "").strip()
    return ""


def _split_resolution(resolution: str, fallback_name: str) -> tuple[str, str]:
    """Parse a placeholder DefaultResolution like 'Name, * (Vendor)' into (version, name)."""
    if not resolution:
        return "", fallback_name
    if "," in resolution:
        name_part, rest = resolution.split(",", 1)
        version_part = rest.strip()
        paren_idx = version_part.find("(")
        version = version_part[:paren_idx].strip() if paren_idx >= 0 else version_part
        return version, name_part.strip()
    return "", resolution.strip()


def parse_plcproj(path: Path) -> dict:
    """Parse a .plcproj for library references.

    The .plcproj uses the MSBuild XML namespace
    (http://schemas.microsoft.com/developer/msbuild/2003). Library refs live
    in <ItemGroup> elements as either <PlaceholderReference> (versioned via a
    DefaultResolution string of the form 'Name, version (Vendor)') or
    <LibraryReference> (Include attribute formatted 'Name,version,Vendor').

    Returns:
        {"libraries": [{"name": str, "version": str, "placeholder": str | None}]}
    """
    root = parse_file(path)
    libraries: list[dict] = []

    for item_group in root:
        if _local(item_group.tag) != "ItemGroup":
            continue
        for ref in item_group:
            tag = _local(ref.tag)
            include = ref.get("Include", "").strip()
            if tag == "PlaceholderReference":
                resolution = _child_text(ref, "DefaultResolution")
                version, resolved_name = _split_resolution(resolution, include)
                libraries.append({
                    "name": resolved_name or include,
                    "version": version,
                    "placeholder": include or None,
                })
            elif tag == "LibraryReference":
                parts = [p.strip() for p in include.split(",")]
                name = parts[0] if parts else include
                version = parts[1] if len(parts) > 1 else ""
                libraries.append({
                    "name": name,
                    "version": version,
                    "placeholder": None,
                })

    return {"libraries": libraries}


def parse_tsproj(path: Path) -> dict:
    """Parse a .tsproj for System Manager task definitions.

    .tsproj CycleTime is in 100ns ticks; converted to microseconds here for
    consistency with .TcTTO. POU bindings live in .TcTTO, not .tsproj, so the
    'programs' list is always empty from this source.

    Returns:
        {"tasks": [{"name": str, "cycle_time_us": int | None,
                    "priority": int | None, "programs": []}]}
    """
    root = parse_file(path)
    tasks: list[dict] = []

    for task_el in root.iter():
        if _local(task_el.tag) != "Task":
            continue
        # System-manager tasks carry numeric Id + CycleTime attributes;
        # filters out unrelated <Task> nodes that may appear elsewhere.
        if task_el.get("Id") is None or task_el.get("CycleTime") is None:
            continue
        name = _child_text(task_el, "Name")
        if not name:
            continue
        cycle_ticks = _to_int(task_el.get("CycleTime"))
        cycle_us = cycle_ticks // 10 if cycle_ticks is not None else None
        priority = _to_int(task_el.get("Priority"))
        tasks.append({
            "name": name,
            "cycle_time_us": cycle_us,
            "priority": priority,
            "programs": [],
        })

    return {"tasks": tasks}


def parse_tctto(path: Path) -> dict:
    """Parse a .TcTTO PLC task object file.

    Authoritative source for PLC task layout: contains the cycle time
    (already in microseconds per Beckhoff's own comment), priority, and the
    POU bound to the task via <PouCall><Name>.

    Returns:
        {"name": str, "cycle_time_us": int | None,
         "priority": int | None, "programs": list[str]}
    """
    root = parse_file(path)
    task_el = root.find("Task")
    if task_el is None:
        raise ValueError(f"No <Task> element found in {path}")

    name = task_el.get("Name", "")
    cycle_us = _to_int(_child_text(task_el, "CycleTime"))
    priority = _to_int(_child_text(task_el, "Priority"))

    programs: list[str] = []
    for child in task_el:
        if _local(child.tag) != "PouCall":
            continue
        pou_name = _child_text(child, "Name")
        if pou_name:
            programs.append(pou_name)

    return {
        "name": name,
        "cycle_time_us": cycle_us,
        "priority": priority,
        "programs": programs,
    }

File: dotnet/src/TcKit.Adapters.Xml/TcFileParser.cs
using System.Text.RegularExpressions;
using System.Xml;
using System.Xml.Linq;
using TcKit.Core.Models;

namespace TcKit.Adapters.Xml;

/// <summary>
/// Metadata extractors for TwinCAT project files (.TcPOU / .TcGVL / .TcDUT / .TcTTO /
/// .tsproj / .plcproj). Ported from the Python <c>tckit/utils/tc_file_parser.py</c>;
/// only the fields <c>get_structure</c> needs are extracted (names, types, tasks,
/// libraries), not declarations or bodies. Stdlib XML only.
/// </summary>
internal static class TcFileParser
{
    internal sealed record PouMeta(string Name, PouType Type);

    internal sealed record DutMeta(string Name, DutKind Kind);

    internal sealed record LibRaw(string Name, string Version, string? Placeholder);

    internal sealed record TaskRaw(string Name, int? CycleTimeUs, int? Priority, IReadOnlyList<string> Programs);

    internal sealed record AccessorPart(string Declaration, string Body);

    internal sealed record MemberPart(string Name, string Declaration, string Body);

    internal sealed record PropertyPart(string Name, string Declaration, AccessorPart? Get, AccessorPart? Set);

    internal sealed record PouFull(
        string Name, PouType Type, string Declaration, string Body,
        IReadOnlyList<MemberPart> Methods, IReadOnlyList<MemberPart> Actions, IReadOnlyList<PropertyPart> Properties);

    internal sealed record GvlFull(string Name, string Declaration);

    internal sealed record DutFull(string Name, string Declaration, DutKind Kind, string BaseType);

    /// <summary>Parse a .TcPOU for its name and POU type. Handles &lt;POU&gt; and &lt;Itf&gt; roots.</summary>
    internal static PouMeta ParsePou(string path)
    {
        var root = Load(path);
        var container = Child(root, "POU") ?? Child(root, "Itf")
            ?? throw new InvalidDataException($"No <POU> or <Itf> element found in {path}");
        var name = container.Attribute("Name")?.Value ?? "";
        var declaration = Declaration(container);
        return new PouMeta(name, DetectPouType(declaration, container.Name.LocalName));
    }

    /// <summary>Parse a .TcGVL for its name.</summary>
    internal static string ParseGvl(string path)
    {
        var root = Load(path);
        var gvl = Child(root, "GVL")
            ?? throw new InvalidDataException($"No <GVL> element found in {path}");
        return gvl.Attribute("Name")?.Value ?? "";
    }

    /// <summary>Parse a .TcPOU in full: declaration, body, methods, actions, and properties (with accessors).</summary>
    internal static PouFull ParsePouFull(string path)
    {
        var root = Load(path);
        var container = Child(root, "POU") ?? Child(root, "Itf")
            ?? throw new InvalidDataException($"No <POU> or <Itf> element found in {path}");
        var name = container.Attribute("Name")?.Value ?? "";
        var declaration = Declaration(container);
        var type = DetectPouType(declaration, container.Name.LocalName);

        var methods = container.Elements().Where(e => e.Name.LocalName == "Method")
            .Select(m => new MemberPart(m.Attribute("Name")?.Value ?? "", Declaration(m), StBody(m)))
            .ToList();
        var actions = container.Elements().Where(e => e.Name.LocalName == "Action")
            .Select(a => new MemberPart(a.Attribute("Name")?.Value ?? "", Declaration(a), StBody(a)))
            .ToList();
        var properties = container.Elements().Where(e => e.Name.LocalName == "Property")
            .Select(p =>
            {
                var get = Child(p, "Get");
                var set = Child(p, "Set");
                return new PropertyPart(
                    p.Attribute("Name")?.Value ?? "",
                    Declaration(p),
                    get is not null ? new AccessorPart(Declaration(get), StBody(get)) : null,
                    set is not null ? new AccessorPart(Declaration(set), StBody(set)) : null);
            })
            .ToList();

        return new PouFull(name, type, declaration, StBody(container), methods, actions, properties);
    }

    /// <summary>Parse a .TcGVL for its name and declaration block.</summary>
    internal static GvlFull ParseGvlFull(string path)
    {
        var root = Load(path);
        var gvl = Child(root, "GVL")
            ?? throw new InvalidDataException($"No <GVL> element found in {path}");
        return new GvlFull(gvl.Attribute("Name")?.Value ?? "", Declaration(gvl));
    }

    /// <summary>Parse a .TcDUT for its name, declaration, kind, and (for aliases) base type.</summary>
    internal static DutFull ParseDutFull(string path)
    {
        var root = Load(path);
        var dut = Child(root, "DUT")
            ?? throw new InvalidDataException($"No <DUT> element found in {path}");
        var declaration = Declaration(dut);
        var (kind, baseType) = ClassifyDutFull(declaration);
        return new DutFull(dut.Attribute("Name")?.Value ?? "", declaration, kind, baseType);
    }

    /// <summary>Parse a .TcDUT for its name and kind discriminator.</summary>
    internal static DutMeta ParseDut(string path)
    {
        var root = Load(path);
        var dut = Child(root, "DUT")
            ?? throw new InvalidDataException($"No <DUT> element found in {path}");
        var name = dut.Attribute("Name")?.Value ?? "";
        return new DutMeta(name, ClassifyDut(Declaration(dut)));
    }

    /// <summary>Parse a .plcproj for its library references.</summary>
    internal static List<LibRaw> ParsePlcproj(string path)
    {
        var root = Load(path);
        var libraries = new List<LibRaw>();
        foreach (var itemGroup in root.Elements().Where(e => e.Name.LocalName == "ItemGroup"))
        {
            foreach (var reference in itemGroup.Elements())
            {
                var tag = reference.Name.LocalName;
                var include = (reference.Attribute("Include")?.Value ?? "").Trim();
                if (tag == "PlaceholderReference")
                {
                    var resolution = ChildText(reference, "DefaultResolution");
                    var (version, resolvedName) = SplitResolution(resolution, include);
                    libraries.Add(new LibRaw(
                        string.IsNullOrEmpty(resolvedName) ? include : resolvedName,
                        version,
                        string.IsNullOrEmpty(include) ? null : include));
                }
                else if (tag == "LibraryReference")
                {
                    var parts = include.Split(',').Select(p => p.Trim()).ToArray();
                    var name = parts.Length > 0 ? parts[0] : include;
                    var version = parts.Length > 1 ? parts[1] : "";
                    libraries.Add(new LibRaw(name, version, null));
                }
            }
        }

        return libraries;
    }

    /// <summary>
    /// Return the project paths referenced by a .sln, as written (usually relative to the
    /// solution directory). Solution-folder entries carry the folder name in the path slot;
    /// callers filter those out with a File.Exists check on the resolved path.
    /// </summary>
    internal static List<string> ParseSlnProjectPaths(string path)
    {
        var result = new List<string>();
        foreach (var line in File.ReadLines(path))
        {
            var match = s_slnProject.Match(line);
            if (match.Success)
            {
                // .sln files always store Windows-style backslash paths; normalise so the
                // resolution also works when the reader runs on a non-Windows host (CI).
                result.Add(match.Groups[1].Value.Replace('\\', Path.DirectorySeparatorChar));
            }
        }

        return result;
    }

    private static readonly Regex s_slnProject = new(
        @"^Project\(""\{[^}]*\}""\)\s*=\s*""[^""]*""\s*,\s*""([^""]+)""",
        RegexOptions.Compiled);

    /// <summary>
    /// Parse a .tsproj for System Manager task definitions. CycleTime is in 100 ns ticks;
    /// converted to microseconds for consistency with .TcTTO. Programs are never bound here.
    /// </summary>
    internal static List<TaskRaw> ParseTsproj(string path)
    {
        var root = Load(path);
        var tasks = new List<TaskRaw>();
        foreach (var task in root.Descendants().Where(e => e.Name.LocalName == "Task"))
        {
            // System-manager tasks carry numeric Id + CycleTime attributes; this filters
            // out unrelated <Task> nodes that may appear elsewhere in the file.
            if (task.Attribute("Id") is null || task.Attribute("CycleTime") is null)
            {
                continue;
            }

            var name = ChildText(task, "Name");
            if (string.IsNullOrEmpty(name))
            {
                continue;
            }

            var cycleTicks = ToInt(task.Attribute("CycleTime")?.Value);
            var cycleUs = cycleTicks is not null ? cycleTicks / 10 : null;
            tasks.Add(new TaskRaw(name, cycleUs, ToInt(task.Attribute("Priority")?.Value), []));
        }

        return tasks;
    }

    /// <summary>
    /// Parse a .TcTTO PLC task object: the authoritative source for cycle time (already in
    /// microseconds), priority, and the POU bound to the task via &lt;PouCall&gt;&lt;Name&gt;.
    /// </summary>
    internal static TaskRaw ParseTctto(string path)
    {
        var root = Load(path);
        var task = Child(root, "Task")
            ?? throw new InvalidDataException($"No <Task> element found in {path}");
        var name = task.Attribute("Name")?.Value ?? "";
        var cycleUs = ToInt(ChildText(task, "CycleTime"));
        var priority = ToInt(ChildText(task, "Priority"));

        var programs = new List<string>();
        foreach (var pouCall in task.Elements().Where(e => e.Name.LocalName == "PouCall"))
        {
            var pouName = ChildText(pouCall, "Name");
            if (!string.IsNullOrEmpty(pouName))
            {
                programs.Add(pouName);
            }
        }

        return new TaskRaw(name, cycleUs, priority, programs);
    }

    /// <summary>
    /// Detect the POU type from the element tag and declaration text. &lt;Itf&gt; is always
    /// an interface; otherwise the first keyword in the declaration wins. FUNCTION_BLOCK must
    /// be tested before FUNCTION because it contains it.
    /// </summary>
    internal static PouType DetectPouType(string declaration, string elementTag)
    {
        if (string.Equals(elementTag, "Itf", StringComparison.OrdinalIgnoreCase))
        {
            return PouType.Interface;
        }

        var text = declaration.ToUpperInvariant();
        if (text.Contains("FUNCTION_BLOCK", StringComparison.Ordinal))
        {
            return PouType.FunctionBlock;
        }

        if (text.Contains("FUNCTION", StringComparison.Ordinal))
        {
            return PouType.Function;
        }

        if (text.Contains("PROGRAM", StringComparison.Ordinal))
        {
            return PouType.Program;
        }

        if (text.Contains("INTERFACE", StringComparison.Ordinal))
        {
            return PouType.Interface;
        }

        return PouType.FunctionBlock;
    }

    private static readonly Regex s_blockComment = new(@"\(\*[\s\S]*?\*\)", RegexOptions.Compiled);
    private static readonly Regex s_lineComment = new(@"//[^\r\n]*", RegexOptions.Compiled);
    private static readonly Regex s_pragma = new(@"\{[^}]*\}", RegexOptions.Compiled);
    private static readonly Regex s_typeBody = new(
        @"TYPE\s+[A-Za-z_][A-Za-z0-9_]*\s*(?:EXTENDS\s+[A-Za-z_][A-Za-z0-9_.]*\s*)?:\s*(.+?)(?:END_TYPE|$)",
        RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);

    /// <summary>Discriminate a DUT declaration into its kind (base type discarded).</summary>
    internal static DutKind ClassifyDut(string declaration) => ClassifyDutFull(declaration).Kind;

    /// <summary>
    /// Discriminate a DUT declaration into (kind, baseType). Comments and pragmas are stripped
    /// first so annotated DUTs classify correctly. baseType is the aliased type for ALIAS DUTs and
    /// empty otherwise. Falls back to STRUCT when unparseable.
    /// </summary>
    internal static (DutKind Kind, string BaseType) ClassifyDutFull(string declaration)
    {
        var stripped = s_pragma.Replace(s_lineComment.Replace(s_blockComment.Replace(declaration, " "), " "), " ");
        var match = s_typeBody.Match(stripped);
        if (!match.Success)
        {
            return (DutKind.Struct, "");
        }

        var body = match.Groups[1].Value.Trim();
        var upper = body.ToUpperInvariant();
        if (upper.StartsWith("STRUCT", StringComparison.Ordinal))
        {
            return (DutKind.Struct, "");
        }

        if (upper.StartsWith("UNION", StringComparison.Ordinal))
        {
            return (DutKind.Union, "");
        }

        if (body.StartsWith('('))
        {
            return (DutKind.Enum, "");
        }

        // ALIAS: the body is a type expression terminated by ';'.
        var semi = body.IndexOf(';', StringComparison.Ordinal);
        var baseType = semi >= 0 ? body[..semi].Trim() : body.Trim();
        return (DutKind.Alias, baseType);
    }

    private const string AccessModifiers = @"(?:(?:PUBLIC|PRIVATE|PROTECTED|INTERNAL|FINAL|ABSTRACT)\s+)*";
    private static readonly Regex s_methodReturn = new(
        $@"METHOD\s+{AccessModifiers}\w+\s*:\s*(\w+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex s_propertyReturn = new(
        $@"PROPERTY\s+{AccessModifiers}\w+\s*:\s*(\w+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    // Implementation-only VAR blocks (locals / VAR CONSTANT / VAR_TEMP). The opener must be the
    // only token on its line so VAR_INPUT/OUTPUT/IN_OUT/INST (the API surface) are preserved.
    private static readonly Regex s_localVarBlock = new(
        @"^[ \t]*(?:VAR(?:[ \t]+CONSTANT)?|VAR_TEMP)[ \t]*\r?\n.*?^[ \t]*END_VAR[ \t]*\r?\n?",
        RegexOptions.Multiline | RegexOptions.Singleline | RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>Extract the return type from a METHOD declaration, or "" if absent.</summary>
    internal static string ExtractMethodReturnType(string declaration)
    {
        var match = s_methodReturn.Match(declaration);
        return match.Success ? match.Groups[1].Value : "";
    }

    /// <summary>Extract the return type from a PROPERTY declaration, or "" if absent.</summary>
    internal static string ExtractPropertyReturnType(string declaration)
    {
        var match = s_propertyReturn.Match(declaration);
        return match.Success ? match.Groups[1].Value : "";
    }

    /// <summary>
    /// Strip implementation-only VAR blocks (locals, VAR_TEMP, VAR CONSTANT) from a method
    /// declaration while preserving VAR_INPUT/OUTPUT/IN_OUT/INST. Used for interface signatures;
    /// get_pou_item keeps the full declaration.
    /// </summary>
    internal static string StripMethodLocals(string declaration)
        => s_localVarBlock.Replace(declaration, "").TrimEnd();

    /// <summary>Split a placeholder DefaultResolution like "Name, * (Vendor)" into (version, name).</summary>
    internal static (string Version, string Name) SplitResolution(string resolution, string fallbackName)
    {
        if (string.IsNullOrEmpty(resolution))
        {
            return ("", fallbackName);
        }

        var comma = resolution.IndexOf(',', StringComparison.Ordinal);
        if (comma >= 0)
        {
            var namePart = resolution[..comma];
            var rest = resolution[(comma + 1)..].Trim();
            var paren = rest.IndexOf('(', StringComparison.Ordinal);
            var version = paren >= 0 ? rest[..paren].Trim() : rest;
            return (version, namePart.Trim());
        }

        return ("", resolution.Trim());
    }

    private static XElement Load(string path)
    {
        try
        {
            return XDocument.Load(path).Root
                ?? throw new InvalidDataException($"Empty XML document: {path}");
        }
        catch (XmlException exc)
        {
            throw new InvalidDataException($"XML parse error in {path}: {exc.Message}", exc);
        }
    }

    private static XElement? Child(XElement parent, string localName)
        => parent.Elements().FirstOrDefault(e => e.Name.LocalName == localName);

    private static string Declaration(XElement element)
        => (Child(element, "Declaration")?.Value ?? "").Trim();

    private static string StBody(XElement element)
    {
        var implementation = Child(element, "Implementation");
        return implementation is null ? "" : (Child(implementation, "ST")?.Value ?? "").Trim();
    }

    private static string ChildText(XElement element, string localName)
        => (Child(element, localName)?.Value ?? "").Trim();

    private static int? ToInt(string? value)
        => int.TryParse(value?.Trim(), out var result) ? result : null;
}
